plugins/catalog-backend/src/ingestion/processors/bitbucket/client.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2021 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import fetch from 'cross-fetch';
+
+import {
+  BitbucketIntegrationConfig,
+  getBitbucketRequestOptions,
+} from '@backstage/integration';
+
+export class BitbucketClient {
+  private readonly config: BitbucketIntegrationConfig;
+
+  constructor(options: { config: BitbucketIntegrationConfig }) {
+    this.config = options.config;
+  }
+
+  async listProjects(options?: ListOptions): Promise<PagedResponse<any>> {
+    return this.pagedRequest(`${this.config.apiBaseUrl}/projects`, options);
+  }
+
+  async listRepositories(
+    projectKey: string,
+    options?: ListOptions,
+  ): Promise<PagedResponse<any>> {
+    return this.pagedRequest(
+      `${this.config.apiBaseUrl}/projects/${projectKey}/repos`,
+      options,
+    );
+  }
+
+  private async pagedRequest(
+    endpoint: string,
+    options?: ListOptions,
+  ): Promise<PagedResponse<any>> {
+    const request = new URL(endpoint);
+    if (options) {
+      (Object.keys(options) as Array<keyof typeof options>).forEach(key => {
+        const value: any = options[key] as any;
+        if (value) {
+          request.searchParams.append(key, value);
+        }
+      });
+    }
```

Comment:
uh just do `if (options?.limit !== undefined) { request.searchParams.append('limit', options.limit); }` etc :) this type trickery doesn't feel worth it :P

Reply:
Or iterate using `for (... in ...)` to avoid all the casting 😅

Reply:
Thanks for the advice, I went with @Rugvip 's suggestion